### ISSUE TITLE
[tt_transformers] Garbage outputs for batch != 1 when putting whole prefill inside trace

### ISSUE THREAD

**Issue by nostojicTT** [tenstorrent/tt-metal#35100]
## This bug was seen on Llama-3.2-1B-Instruct so please for debugging use this model and this branch - https://github.com/tenstorrent/tt-metal/tree/nostojic/slicing_tracing
 
- When putting the whole trace processing under trace, the first batch has good outputs, but the next batches have garbage outputs
- I tried couple of setups in order to narrow down where the issue can be:
1. [prefill trace - on, decode trace - on, sampling on device - false] When device sampling is off (set `device_sampling_params = None` in `simple_text_demo.py`) - good outputs 🟢 
2. [prefill trace - on, decode trace - on, internal sampling trace - off] - bad outputs 🔴 
3. [prefill trace - on, decode trace (decode + sampling) - off] - bad outputs 🔴 

- Conclusion based of these 3: Whenever I have tracing in prefill and sampling on device (whether it's trace is on or off, I get bad outputs)

4. [prefill trace - off, decode trace (decode + sampling) -on ] - good outputs 🟢 

### Important
- When warming up decode first, the outputs for the batches after the first one are good 🟢  (this test case is both tested through vLLM and locally)
- Because of this, I am assuming some kind of corruption is happening for the sampling module (since when sampling on device is off this issue does not happen, thats why I think something inside the sampling module is corrupted) between batches (prefill x 32 -> (decode + sampling) x 32 -> prefill (corrupts some locations) x32 -> (decode + sampling) x 32 (bad outputs) )
- Currently I am debugging to see which module (decode or sampling) is creating a difference between batches

**Comment by nostojicTT** [comment]
https://github.com/tenstorrent/tt-metal/blob/d24b55a2168bad9687b9b22a0c84d4344efb5d46/models/common/sampling/tt_sampling.py#L272

After executing this the output is different for the first iteration of the first batch and the first iteration of the second batch. The variable `topk_values` is the same. The way I got to this is by saving the `topk_values` from the iteration 0 of batch0, and later comparing it with the `topk_values` of iteration 0 of batch1. They are the same, but `topk_values_gathered` are different

**Comment by nostojicTT** [comment]
It seems that the issue is the following:

```python3
a = ttnn.slice(b, slice_tensors[0], slice_tensors[1], slice_dim=2, num_devices=x.shape[2] // 32)
b = self.norm(c, mode=mode)
c = ttnn.interleaved_to_sharded(d, self.model_config["LM_HEAD_INPUT_MEMCFG"])
d = self.lm_head(e)
e = ttnn.to_layout(d, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
```

All of these listed operators (slice, norm, interleaved_to_sharded, lm_head and to_layout) create new device buffers (tensors).
Before I get into explaining, I need to point this out:
1. If during trace capture, a certain op, lets say `ttnn.slice`, creates a tensor on address A, when `execute_trace` is called and that same op has to be executed, it will create the buffer on the same address (A) as it was created during capturing the trace.
Here is an example:
- Trace is being captured -> ttnn.slice produces a tensor on address A
- Trace is called -> when we hit that same ttnn.slice is will produce a tensor on address A
2. When trace is finished executing, tensors to which there is no more reference to from python are being deallocated - in this case showed above, tensors `a, b, c, d and e` will be deallocated and the memory that it occupied will be freed

Back to the explanation - lets see what happens when we run our model through vLLM:
Trace prefill warmup (Allocates tensors a,b,c,d,e and when done frees that memory) -> Trace decode warmup (uses the memory addresses that fall into the range where tensors a,b,c,d and e were and creates tensors that are in DRAM on in L1 that should persist from now on) -> Execute Prefill Trace (this trace will corrupt the decode/sampling) -> Execute Decode Trace (works with corrupt memory) -> garbage outputs

Main reason this happens: Decode trace capture sees memory addresses that were occupied by tensors a,b,c,d and e as free so it allocates its tensors there. Solution to this is to keep these tensors always allocated so that the memory allocator does not see them as free and then use them for decode trace warmup.

One solution to this is to create the persistent buffer references in the init and always keep them throughout the whole model execution:
```python3
self.persistent_buffer_a = None
self.persistent_buffer_b = None
self.persistent_buffer_c = None
self.persistent_buffer_d = None
self.persistent_buffer_e = None
```

```python3
self.persistent_buffer_a = ttnn.slice(x, (0, 0, 96, 0), (1, 1, 96 + 32, x.shape[-1]))
self.persistent_buffer_b = self.norm(self.persistent_buffer_a, mode=mode)
self.persistent_buffer_c = ttnn.interleaved_to_sharded(self.persistent_buffer_b, self.model_config["LM_HEAD_INPUT_MEMCFG"])
self.persistent_buffer_d = self.lm_head(self.persistent_buffer_c)
self.persistent_buffer_e = ttnn.to_layout(self.persistent_buffer_d, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
```

What this will do is it will keep the reference to those created tensors the whole model lifecycle and then decode trace capture won't use those addresses when it's trace capturing is being done.

The problem with bad outputs was already experienced with Llama70B and what they did is they implemented persistent buffers for CCL operations. We dont have that currently inside tt_transformers and that can cause issues in the future.
https://github.com/tenstorrent/tt-metal/issues/26423
https://github.com/tenstorrent/tt-metal/pull/27018

**Comment by nostojicTT** [comment]
### Edit  - comment not correct

The persistent buffers for the CCL's are not needed if the problem does not occur, since CCLs need either persistent buffers or barrier semaphores for synchronization. In our case, we have barrier semaphores implemented.

----



CCL persistent buffers problem in tt_transformer (which can cause bad outputs since it can corrupt memory in all batches except for the first one):
```python3
x = layer(
    x,
    current_pos,
    rot_mats_global=rot_mats_global,
    rot_mats_local=rot_mats_local,
    user_id=user_id,
    mode=mode,
    page_table=page_table,
    chunk_page_table=chunk_page_table,
    chunk_start_idx=chunk_start_idx,
    kv_cache=kv_cache[i] if kv_cache is not None else None,
)
```

After each layer is executed, `x.buffer_address()` will be a different value, which means that the CCL operations create new tensors. To avoid this, persistent buffers were added to CCL operations and `x.buffer_address()` is always the same for Llama70B implementation.
```python3
for i, layer in enumerate(self.layers):
    x, h = layer(
        x,
        h,
        current_pos,
        rot_mats,
        user_id,
        mode,
        page_table,
        chunk_page_table=chunk_page_table,
        chunk_start_idx=chunk_start_idx,
        kv_cache=kv_cache[i] if kv_cache is not None else None,
        batch_size=batch_size,
    )
```

`x.buffer_address` will be the same after every layer is executed.

**Comment by nostojicTT** [comment]
```python3
def _perform_all_gather(self, tensor, dim, cluster_axis, memory_config, num_links, buffer_key=None, dtype=None):
    """Flexible all-gather that works with both CCL implementations."""
    if self.cluster_shape[0] * self.cluster_shape[1] == 32:
        # Use line_all_gather with persistent buffer support
        return self.tt_ccl.line_all_gather(
            tensor,
            dim=dim,
            cluster_axis=cluster_axis,
            memory_config=memory_config,
            num_links=num_links,
            buffer_key=buffer_key,
        )
    else:
        # Use tt_all_gather
        cluster_axis = None
        num_links = 1
        tt_logits = ttnn.all_gather(
            tensor,
            dim=dim,
            num_links=num_links,
            memory_config=tensor.memory_config(),
            cluster_axis=cluster_axis,
            topology=ttnn.Topology.Linear,
        )
        return tt_logits
```

When changing the CCL to use the `ttnn.experimental.all_gather_async`, the problems go away.

**Comment by nostojicTT** [comment]
Update - Can't create a minimal repro test so I created the following test:
1. Create the tensor that will be the input tensor for the all_gather
2. Run inside a loop the following : (prefill -> all_gather) x n

The results of the test are:
1. Trace prefill disabled - all outputs of the all_gather are good
2. Trace prefill enabled - first output of all_gather is good, all subsequent are bad
3. Trace prefill enabled, barrier semaphore passed to CCL (experimental called) - good outputs for all all_gathers
4. Trace prefill enabled, persistent buffer passed to CCL - good outputs for all all_gathers
 
4. (important) After the step 1. I ran all_gather once before the first prefill. When I did that, all outputs of the all_gather are good? -> something gets corrupted, could be program cache?
